### ISSUE TITLE
Triggers

### ISSUE THREAD

**Issue by alexanderdean** [snowplow/factotum#101]
This is the idea of using a poll-based trigger to kick off a factfile. This is a potential building block for reactive batch.

The process would be as follows:

* Every poll interval, execute a "trigger process"
* If the trigger process returns 0, then proceed to execute the factfile
* If the trigger process returns non-0, then exit without running the factfile
* Emit an appropriate webhook in both cases

The nice thing about this approach is that you can write your triggers in anything - Factotum doesn't need to know about your API requests or `_SUCCESS` file checks; if it can be run in a shell, then it can be used as a job trigger.

A couple of outstanding questions:

1. Whether we support multiple triggers and boolean operations to combine them
2. Where the scheduling of the polling should live (e.g. inside a long-running Factotum daemon or external to Factotum, in a cron)

**Comment by alexanderdean** [comment]
Thoughts @ninjabear @jbeemster @ungn ?

**Comment by ninjabear** [comment]
I think for "on demand" jobs the triggers could vary a lot (poll periods and so on). A good interface would be to allow third party apps to trigger a job by POSTing to /runs (or similar route).

**Comment by alexanderdean** [comment]
That's a really interesting idea @ninjabear - so we have a **push trigger** (REST / RPC against a daemon) versus a **pull trigger** (regularly running a test to see if a condition is met)?

**Comment by alexanderdean** [comment]
You could argue that our internal Snowdroid behavior is already a form of **push trigger**?

**Comment by ninjabear** [comment]
Yep - exactly my thoughts, this is how snowdroid works today. We'd essentially be unifying cron and the executor server.

Also pull triggers are a bit problematic security wise - push style triggers can be hosted anywhere (they can be provided an access token)

**Comment by alexanderdean** [comment]
I don't think you can get rid of a pull trigger altogether - all you can do is move it further outside the stack (so there's an external polling check that when it passes activates a push trigger). At the moment we have de facto pull triggers inside of EmrEtlRunner, insofar as we kick off the job regularly and don't proceed (`if in_bucket is empty or enriched_bucket is full or shredded_bucket is full`). So we need to move that logic further out, and a formal pull trigger mechanism somewhere seems like one approach.

> Also pull triggers are a bit problematic security wise

Can you elaborate on the security risks @ninjabear ?

**Comment by ninjabear** [comment]
Can you elaborate on the security risks?

A *pull style trigger* executes arbitrary code on the cluster that has access to job information (and so on). A *push style trigger* can run somewhere else (it means if desired, we could let third parties write their own triggers).

It's not really a dealbreaker, it could be locked down either way. My point was really that pull style triggers need a bit more thought in this regard! 

    I don't think you can get rid of a pull trigger altogether - all you can do is move it further outside the stack (so e.g. there's an external polling check that, when it passes, activates a push trigger)

I agree! What we're really doing is allowing pull style triggers through push style triggers. If you wanted to get really meta you could have another scheduled job that hits the push trigger of another job if a condition is met (using NOOPs).

**Comment by alexanderdean** [comment]
Agree, agree. Exciting!
